Add `LeaveOpen` parameter to StreamReader/Writer constructors

ELLION911 opened the pull request:
## Description
- Added a parameter LeaveOpen for stream reader/writer.
- Added a unit test for it.

## Motivation and Context
Sometime need to use stream writer/reader for a short period of time, and then continue to work with stream without disposing original stream. Current patch allow to do it.

## How Has This Been Tested?

- We created a test app, and write some buffer to stream by stream writer. 
- Release Writer, but not a original stream
- Use stream reader with same stream couple times.
- We also added unit test project with two tests cases.

## Screenshots

## Types of changes
- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [x] Unit Tests (add new Unit Test(s) or improved existing one(s), has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist:
- [x] My code follows the code style of this project (only if there are changes in source code).
- [x] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [ ] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).
- [x] I have added new tests to cover my changes.


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **Chores**
  - Added a lock file to ensure consistent package versions for the unit test project.
  - Updated the build pipeline to enable and configure unit test execution during library builds.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
Add LeaveOpen Parameter for Stream Reader/Writer and unit tests

Eugene-SvitSolutions commented:
> @dotnet-policy-service agree [company="ShockTracker"]

Eugene-SvitSolutions commented:
@dotnet-policy-service agree

Eugene-SvitSolutions commented:
@dotnet-policy-service agree company="Microsoft"

josesimoes commented on System.IO.Streams/StreamReader.cs:
```diff
@@ -59,14 +60,19 @@ public bool EndOfStream
+        }
 
         /// <summary>
         /// Initializes a new instance of the <see cref="StreamReader"/> class for the specified stream.
```
```suggestion
        /// Initializes a new instance of the <see cref="StreamReader"/> class for the specified stream and optionally leaves the stream open.
```

josesimoes commented on System.IO.Streams/StreamWriter.cs:
```diff
@@ -51,9 +57,10 @@ public override string NewLine
         /// Initializes a new instance of the <see cref="StreamWriter"/> class for the specified stream by using UTF-8 encoding and the default buffer size.
```
```suggestion
        /// Initializes a new instance of the <see cref="StreamWriter"/> class for the specified stream by using UTF-8 encoding and the default buffer size and optionally leaves the stream open.
```

josesimoes commented on System.IO.Streams/StreamWriter.cs:
```diff
@@ -51,9 +57,10 @@ public override string NewLine
         /// Initializes a new instance of the <see cref="StreamWriter"/> class for the specified stream by using UTF-8 encoding and the default buffer size.
         /// </summary>
         /// <param name="stream">The stream to write to.</param>
+        /// <param name="leaveOpen">Leave stream open after dispose writer.</param>
```
```suggestion
        /// <param name="leaveOpen"><see langword="true"/> to leave the stream open after the <see cref="StreamWriter"/> object is disposed; otherwise, <see langword="false"/>.</param>
```

josesimoes commented on System.IO.Streams/StreamReader.cs:
```diff
@@ -59,14 +60,19 @@ public bool EndOfStream
         /// Initializes a new instance of the <see cref="StreamReader"/> class for the specified stream.
         /// </summary>
         /// <param name="stream">The stream to be read.</param>
+        /// <param name="leaveOpen">Leave stream open after dispose writer.</param>
```
```suggestion
                /// <param name="leaveOpen"><see langword="true"/> to leave the stream open after the <see cref="StreamReader"/> object is disposed; otherwise, <see langword="false"/>.</param>
```

josesimoes reviewed (commented):
All good, just some minor changes on the IntelliSense comments (from .NET API docs).

ELLION911 committed:
Update System.IO.Streams/StreamReader.cs

Co-authored-by: José Simões <jose.simoes@eclo.solutions>

ELLION911 committed:
Update System.IO.Streams/StreamWriter.cs

Co-authored-by: José Simões <jose.simoes@eclo.solutions>

ELLION911 committed:
Update System.IO.Streams/StreamWriter.cs

Co-authored-by: José Simões <jose.simoes@eclo.solutions>

ELLION911 committed:
Update System.IO.Streams/StreamReader.cs

Co-authored-by: José Simões <jose.simoes@eclo.solutions>

josesimoes committed:
Enable unit tests in AZSO pipeline

josesimoes commented:
/AzurePipelines run

josesimoes committed:
remove runsettings file in test project

josesimoes commented:
/AzurePipelines run

josesimoes commented:
@Eugene-SvitSolutions there are two unit tests failing in the pipeline. Care to check please?

Commit:
Fix Unit test error

ELLION911 reviewed (commented):
Code fixed

josesimoes commented on UnitTests/StreamReaderWriterUnitTests/Writer.cs:
```diff
@@ -0,0 +1,69 @@
+
+            using (var stream = new MemoryStream())
+            {
+                //stream.SetLength(0);
```
```suggestion
```

josesimoes committed:
Remove runsettings from test projects

josesimoes committed:
Update general runsettings file

josesimoes committed:
Work on new unit tests

- Rename unit tests files for clarity.
- Remove unused usings.
- Rename locals to follow project code style.
- Migrate using to modern pattern.
- Replace try/catch with Assert for object disposed exception.
- Minor code style fixes.

josesimoes reviewed (approved):
LGTM! 
I've made a final pass to fix code style and replace try/catch with Assert on exception.
Thank you for working on this. 👍🏻

josesimoes committed:
Update test framework
